lms/djangoapps/certificates/models.py

```diff
@@ -547,3 +549,39 @@ def get_config(cls):
         instance = cls.current()
         json_data = json.loads(instance.configuration) if instance.enabled else {}
         return json_data
+
+
+class BadgeTracker(models.Model):
+    """
+    Tracks badges on our side of the badge baking transaction
+    """
+    user = models.ForeignKey(User)
+    course_id = CourseKeyField(max_length=255, blank=True, default=None)
+    data = JSONField()
+
+    def image_url(self):
+        return self.data['image']
+
+    class Meta(object):
+        unique_together = (('course_id', 'user'),)
+
+
+# Create badges on certificate save.
+def create_badge(sender, instance, **kwargs):
+    if not settings.FEATURES['ENABLE_OPENBADGES']:
+        return
+    if not modulestore().get_course(instance.course_id).issue_badges:
+        return
+    if BadgeTracker.objects.filter(user=instance.user, course_id=instance.course_id):
+        # Badge already exists. Skip.
```

Comment:
@Kelketek might make sense to add some logging here and on line 574: these are special situations so it might help debugging it someday.